From 14f3669dd74312923ce442b6fea9afca9651dafb Mon Sep 17 00:00:00 2001
From: Sultan Alsawaf <sultan@kerneltoast.com>
Date: Tue, 26 Aug 2025 20:00:42 -0700
Subject: [PATCH] ACPI: processor: Disable bus master check for AMD

The bus master check on AMD is observed (at least on both Strix Point and
Strix Halo) to be very costly and results in not only increased power
consumption, but also lots of CPU cycles burned. Disable the actual bus
master check while keeping pr->flags.bm_check enabled so there's no cache
flush upon C3 entry.

Signed-off-by: Sultan Alsawaf <sultan@kerneltoast.com>
Signed-off-by: Eric Naim <dnaim@cachyos.org>
---
 drivers/acpi/processor_idle.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/drivers/acpi/processor_idle.c b/drivers/acpi/processor_idle.c
index 4482cf28f56a..9c93e8858024 100644
--- a/drivers/acpi/processor_idle.c
+++ b/drivers/acpi/processor_idle.c
@@ -1427,6 +1427,17 @@ void acpi_processor_power_init(struct acpi_processor *pr)
 	if (disabled_by_idle_boot_param())
 		return;
 
+	/*
+	 * The bus master check on AMD is observed to be very costly and results
+	 * in not only increased power consumption, but also lots of CPU cycles
+	 * burned. Disable the actual bus master check while keeping
+	 * pr->flags.bm_check enabled so there's no cache flush upon C3 entry.
+	 */
+	if (!bm_check_disable && boot_cpu_data.x86_vendor == X86_VENDOR_AMD) {
+		bm_check_disable = true;
+		pr_info("Enabled processor.bm_check_disable for AMD cpuidle\n");
+	}
+
 	if (!acpi_processor_get_power_info(pr))
 		pr->flags.power_setup_done = 1;
 
-- 
2.55.0

